File: 3.x.x/crates-utopia/auth/src/store.rs
//! Base64-encodable key/value authentication state store.

use std::collections::HashMap;

use base64::engine::general_purpose::STANDARD;
use base64::Engine;
use serde_json::Value;

use crate::error::AuthError;

/// Serializable property map for authentication state.
#[derive(Debug, Clone, Default)]
pub struct Store {
    data: HashMap<String, Value>,
    key: Option<String>,
}

impl Store {
    /// Create an empty store.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Read a property value.
    #[must_use]
    pub fn get_property(&self, key: &str) -> Option<&Value> {
        self.data.get(key)
    }

    /// Set a property value.
    pub fn set_property(&mut self, key: impl Into<String>, value: impl Into<Value>) -> &mut Self {
        self.data.insert(key.into(), value.into());
        self
    }

    /// Read the store encryption key, if any.
    #[must_use]
    pub fn key(&self) -> Option<&str> {
        self.key.as_deref()
    }

    /// Set the store encryption key.
    pub fn set_key(&mut self, key: Option<impl Into<String>>) -> &mut Self {
        self.key = key.map(Into::into);
        self
    }

    /// Encode store data as a base64 JSON string.
    pub fn encode(&self) -> Result<String, AuthError> {
        let json = serde_json::to_string(&self.data)?;
        Ok(STANDARD.encode(json))
    }

    /// Decode a base64 JSON string into this store.
    ///
    /// Invalid input is ignored and leaves the store unchanged (matching PHP behavior).
    pub fn decode(&mut self, data: &str) -> &mut Self {
        let Ok(decoded) = STANDARD.decode(data) else {
            return self;
        };

        let Ok(json) = serde_json::from_slice::<Value>(&decoded) else {
            return self;
        };

        let Some(object) = json.as_object() else {
            return self;
        };

        for (key, value) in object {
            self.set_property(key.clone(), value.clone());
        }

        self
    }

    /// All stored properties.
    #[must_use]
    pub fn properties(&self) -> &HashMap<String, Value> {
        &self.data
    }
}

File: packages/auth/src/Auth/Store.php
<?php

namespace Utopia\Auth;

class Store
{
    /**
     * @var array<string, mixed>
     */
    protected array $data = [];

    protected ?string $key = null;

    /**
     * Get a property from the store
     */
    public function getProperty(string $key, mixed $default = null): mixed
    {
        return $this->data[$key] ?? $default;
    }

    /**
     * Set a property in the store
     */
    public function setProperty(string $key, mixed $value): static
    {
        $this->data[$key] = $value;

        return $this;
    }

    /**
     * Get the store key
     */
    public function getKey(): ?string
    {
        return $this->key;
    }

    /**
     * Set the store key
     */
    public function setKey(?string $key): static
    {
        $this->key = $key;

        return $this;
    }

    /**
     * Encode store data to base64 string
     *
     *
     * @throws \JsonException
     */
    public function encode(): string
    {
        $json = json_encode($this->data, JSON_THROW_ON_ERROR);

        return base64_encode($json);
    }

    /**
     * Decode base64 string and populate current store instance
     */
    public function decode(string $data): static
    {
        try {
            $decoded = base64_decode($data, true);
            if ($decoded === false) {
                return $this;
            }

            $json = json_decode($decoded, true, 512, JSON_THROW_ON_ERROR);
            if (\is_array($json)) {
                foreach ($json as $key => $value) {
                    $this->setProperty($key, $value);
                }
            }
        } catch (\JsonException) {
            // Invalid JSON, return empty store
        }

        return $this;
    }
}
